logd/store/badger/history.go

```diff
@@ -2,66 +2,120 @@ package badger
 
 import (
 	"context"
-	"encoding/binary"
 	"fmt"
-	"strings"
 
-	"github.com/deref/exo/chrono"
 	"github.com/deref/exo/logd/api"
+	"github.com/deref/exo/logd/store"
+	"github.com/deref/exo/util/binaryutil"
 	"github.com/dgraph-io/badger/v3"
-	"github.com/oklog/ulid/v2"
 )
 
-func (log *Log) GetLastEventAt(ctx context.Context) *string {
+func (log *Log) GetLastEvent(ctx context.Context) *api.Event {
+	lastEvent, err := log.getLastEvent(ctx)
+	if err != nil || lastEvent == nil {
+		return nil
+	}
+
+	return lastEvent
+}
+
+func (log *Log) getLastEvent(ctx context.Context) (*api.Event, error) {
+	var event *api.Event
 	prefix := append([]byte(log.name), 0)
-	var timestamp uint64
-	err := log.db.View(func(txn *badger.Txn) error {
+
+	if err := log.db.View(func(txn *badger.Txn) error {
 		opts := badger.DefaultIteratorOptions
 		opts.PrefetchSize = 1
 		opts.Reverse = true
+
 		it := txn.NewIterator(opts)
 		defer it.Close()
+
 		it.Seek(append([]byte(log.name), 255))
 		if it.ValidForPrefix(prefix) {
 			item := it.Item()
+			key := item.Key()
 			return item.Value(func(val []byte) error {
 				if err := validateVersion(val[versionOffset]); err != nil {
 					return err
 				}
-				timestamp = binary.BigEndian.Uint64(val[timestampOffset : timestampOffset+timestampLen])
+				lastEvent, err := eventFromEntry(log.name, key, val)
+				if err != nil {
+					return err
+				}
+				event = &lastEvent
 				return nil
 			})
 		}
 		return nil
-	})
-	if err != nil {
-		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("scanning log: %w", err)
 	}
 
-	if timestamp != 0 {
-		s := chrono.NanoToIso(int64(timestamp))
-		return &s
-	}
-	return nil
+	return event, nil
 }
 
-func (log *Log) GetEvents(ctx context.Context, after string, limit int) ([]api.Event, error) {
-	events := make([]api.Event, 0, limit)
-	if err := log.db.View(func(txn *badger.Txn) error {
-		it := txn.NewIterator(badger.DefaultIteratorOptions)
+func (log *Log) GetLastCursor(ctx context.Context) (cursor *store.Cursor) {
+	cursor = new(store.Cursor)
+	prefix := append([]byte(log.name), 0)
+
+	_ = log.db.View(func(txn *badger.Txn) error {
```

Comment:
i _think_ that GetLastCursor no longer should swallow errors, since the call site now does return an error.